File: src/invoiceops_agent/api/schemas/provenance.py
"""Auditor-facing event trace and cross-run invoice provenance contracts."""

from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from invoiceops_agent.api.schemas.invoice_read import InvoiceSource, InvoiceStatus, RunStatus
from invoiceops_agent.ledger.schemas import (
    ActorType,
    InvoiceCursor,
    LedgerEvent,
    RunCursor,
    TraceId,
    UtcDatetime,
    Version,
    VersionPins,
)


class ProvenanceModel(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True, allow_inf_nan=False)


class TraceEvent(ProvenanceModel):
    id: UUID
    sequence: int = Field(ge=1)
    event_type: Version
    node: Version | None
    actor_type: ActorType
    actor_id: Version
    supersedes_id: UUID | None
    versions: VersionPins
    created_at: UtcDatetime


class RunTracePage(ProvenanceModel):
    run_id: UUID
    invoice_id: UUID
    trace_id: TraceId
    status: RunStatus
    graph_version: Version
    started_at: UtcDatetime | None
    completed_at: UtcDatetime | None
    events: list[TraceEvent] = Field(max_length=200)
    next_cursor: RunCursor | None


class InvoiceProvenancePage(ProvenanceModel):
    invoice_id: UUID
    status: InvoiceStatus
    source: InvoiceSource
    created_at: UtcDatetime
    events: list[LedgerEvent] = Field(max_length=200)
    next_cursor: InvoiceCursor | None

File: frontend/src/schemas/provenance.ts
import { z } from "zod";
import { invoiceLedgerCursorSchema, ledgerEventSchema, runLedgerCursorSchema } from "./ledger";
import { invoiceSourceSchema, invoiceStatusSchema, runStatusSchema } from "./invoice_read";

export const traceEventSchema = ledgerEventSchema.omit({
  payload: true, run_id: true, invoice_id: true,
});

export const runTracePageSchema = z.object({
  run_id: z.string().uuid(),
  invoice_id: z.string().uuid(),
  trace_id: z.string().regex(/^[0-9a-f]{32}$/),
  status: runStatusSchema,
  graph_version: z.string().min(1).max(128).regex(/\S/),
  started_at: z.string().datetime({ offset: true }).nullable(),
  completed_at: z.string().datetime({ offset: true }).nullable(),
  events: z.array(traceEventSchema).max(200),
  next_cursor: runLedgerCursorSchema.nullable(),
}).strict();

export const invoiceProvenancePageSchema = z.object({
  invoice_id: z.string().uuid(),
  status: invoiceStatusSchema,
  source: invoiceSourceSchema,
  created_at: z.string().datetime({ offset: true }),
  events: z.array(ledgerEventSchema).max(200),
  next_cursor: invoiceLedgerCursorSchema.nullable(),
}).strict();

export type RunTracePage = z.infer<typeof runTracePageSchema>;
export type InvoiceProvenancePage = z.infer<typeof invoiceProvenancePageSchema>;
